text/0000-expand-math.md

```diff
@@ -0,0 +1,230 @@
+FloatingPoint <-- BigFloat
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+## Methods of Concern
+
+```pony
+trait val Real[A: Real[A] val] is
+  (Stringable & _ArithmeticConvertible & Comparable[A])
+  ...
+  new val min_value()
+  new val max_value()
+  ...
+```
+
+`Real` will exist above `Rational`, `BigInt`, and `BigFloat` and as such would require defining the above methods for these types. `Rational` can be defined as minimum and maximum of the numerator, however `BigInt` and `BigFloat` by definition have arbitrary precision making defining a minimum and maximum difficult at the least -- if we define them as the minimum and maximum of a machine-sized int and float, or define them as -Inf and Inf -- or impossible at the worst -- if we define them by their possible limits which are arbitrary.
+
+```pony
+trait val Integer[A: Integer[A] val] is Real[A]
+  ...
+  fun op_and(y: A): A => this and y
+  fun op_or(y: A): A => this or y
+  fun op_xor(y: A): A => this xor y
+  fun op_not(): A => not this
+
+  fun bit_reverse(): A
+    """
+    Reverse the order of the bits within the integer.
+    For example, 0b11101101 (237) would return 0b10110111 (183).
+    """
+
+  fun bswap(): A
+```
+
+`Integer` will exist above `BigInt` and as such would require defining the above methods -- however `BigInt` will be defined via other numerics so these methods could be applied recursively.
+
+```pony
+trait val FloatingPoint[A: FloatingPoint[A] val] is Real[A]
+  new val min_normalised()
+  new val epsilon()
+  fun tag radix(): U8
+  fun tag precision2(): U8
+  fun tag precision10(): U8
+  fun tag min_exp2(): I16
+  fun tag min_exp10(): I16
+  fun tag max_exp2(): I16
+  fun tag max_exp10(): I16
+  ...
+  fun abs(): A
+  fun ceil(): A
+  fun floor(): A
+  fun round(): A
+  fun trunc(): A
+
+  fun finite(): Bool
+  fun infinite(): Bool
+  fun nan(): Bool
+
+  fun ldexp(x: A, exponent: I32): A
+  fun frexp(): (A, U32)
+  fun log(): A
+  fun log2(): A
+  fun log10(): A
+  fun logb(): A
+
+  fun pow(y: A): A
+  fun powi(y: I32): A
+
+  fun sqrt(): A
+
+  fun sqrt_unsafe(): A
+    """
+    Unsafe operation.
+    If this is negative, the result is undefined.
+    """
+
+  fun cbrt(): A
+  fun exp(): A
+  fun exp2(): A
+
+  fun cos(): A
+  fun sin(): A
+  fun tan(): A
+
+  fun cosh(): A
+  fun sinh(): A
+  fun tanh(): A
+
+  fun acos(): A
+  fun asin(): A
+  fun atan(): A
+  fun atan2(y: A): A
+
+  fun acosh(): A
+  fun asinh(): A
+  fun atanh(): A
+```
+
+`FloatingPoint` will exist above `BigFloat` and as such would require defining the above methods -- many of which are ill-defined under arbitrary precision, or are functions using C-FFI and/or LLVM intrinsics.
```

Comment:
I'm confused. Why would we do it this way if it has these problems? This seems like an argument against doing it this way. I feel like there's some meaning here that I am missing.